File: Step 1 - Learn the Basics/1.4 Know Basic Maths/04 GCD Or HCF/euclidean.java
// https://www.geeksforgeeks.org/problems/lcm-and-gcd4516/1

class Solution 
{
    // Method to calculate GCD using Euclidean Algorithm
    public static int gcd(int a, int b)
    {
        while (b != 0)
        {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    // Method to calculate LCM using the formula
    public static int lcm(int a, int b)
    {
        // LCM(a, b) = |a * b| / GCD(a, b)
        return (a * b) / gcd(a, b);
    }

    // Method to calculate both LCM and GCD and return them as an array
    public static int[] lcmAndGcd(int a, int b)
    {
        int[] result = new int[2]; // Array to store LCM and GCD

        // Calculate GCD using Euclidean algorithm
        int gcdValue = gcd(a, b);

        // Calculate LCM
        int lcmValue = lcm(a, b);

        // Store results in the array
        result[0] = lcmValue;  // LCM
        result[1] = gcdValue;  // GCD

        return result;  // Return the result array
    }
}

File: Step 1 - Learn the Basics/1.4 Know Basic Maths/04 GCD Or HCF/euclidean.py
class Solution:
    @staticmethod
    def gcd(a: int, b: int) -> int:
        # Method to calculate GCD using the Euclidean Algorithm
        while b != 0:
            a, b = b, a % b
        return a

    @staticmethod
    def lcm(a: int, b: int) -> int:
        # Method to calculate LCM using the formula
        # LCM(a, b) = |a * b| / GCD(a, b)
        return (a * b) // Solution.gcd(a, b)

    @staticmethod
    def lcm_and_gcd(a: int, b: int) -> list:
        # Calculate GCD using Euclidean algorithm
        gcd_value = Solution.gcd(a, b)

        # Calculate LCM
        lcm_value = Solution.lcm(a, b)

        # Return results as a list [LCM, GCD]
        return [lcm_value, gcd_value]


# Example usage
solution = Solution()
a, b = 12, 15
result = solution.lcm_and_gcd(a, b)
print(f"LCM: {result[0]}, GCD: {result[1]}")
